language/src/main.rs

```diff
@@ -65,69 +69,226 @@ impl HacspecErrorEmitter for Session {
+                                    );
+                                    new_specials
+                                        .enums
+                                        .extend(special_names_map[&krate_use_string].enums.clone());
+                                    new_specials.aliases.extend(
+                                        special_names_map[&krate_use_string].aliases.clone(),
+                                    );
+                                }
+                            }
+                            _ => (),
+                        };
+                    }
+                    _ => (),
                 }
-            };
-        let krate = match typechecker::typecheck_program(&compiler.session(), &krate, &mut top_ctx)
+            }
+
+            // Remove the modules statements from the crate
+            rustc_ast::ast::Crate {
+                attrs,
+                items: items
+                    .clone()
+                    .into_iter()
+                    .filter(|x| match x.kind {
+                        rustc_ast::ast::ItemKind::Mod(
+                            rustc_ast::ast::Unsafe::No,
+                            rustc_ast::ast::ModKind::Unloaded,
+                        ) => false,
+                        _ => true,
+                    })
+                    .collect(),
+                span,
+            }
+        }
+    };
+
+    /////////////////////////////////
+    // Start of actual translation //
+    /////////////////////////////////
+
+    let external_data = |imported_crates: &Vec<rustspec::Spanned<String>>| {
+        queries.global_ctxt().unwrap().peek_mut().enter(|tcx| {
+            hir_to_rustspec::retrieve_external_data(&compiler.session(), &tcx, imported_crates)
+        })
+    };
+
+    let krate =
+        match ast_to_rustspec::translate(&compiler.session(), &krate, &external_data, new_specials)
         {
             Ok(krate) => krate,
             Err(_) => {
                 compiler
                     .session()
-                    .err("found some Hacspec typechecking errors");
+                    .err("unable to translate to Hacspec due to out-of-language errors");
                 return Compilation::Stop;
             }
         };
+
+    let krate = match name_resolution::resolve_crate(
+        &compiler.session(),
+        krate,
+        &external_data,
+        new_top_ctx,
+    ) {
+        Ok(krate) => krate,
+        Err(_) => {
+            compiler
+                .session()
+                .err("found some Hacspec name resolution errors");
+            return Compilation::Stop;
+        }
+    };
+
+    let krate = match typechecker::typecheck_program(&compiler.session(), &krate, new_top_ctx) {
+        Ok(krate) => krate,
+        Err(_) => {
+            compiler
+                .session()
+                .err("found some Hacspec typechecking errors");
+            return Compilation::Stop;
+        }
+    };
+
+    if root_module {
```

Comment:
This statement might be incorrect, since it no longer checks code in submodules. There is not really a point where we have just typechecked the modules / crate, but not yet made any files. I am not sure whether we should change the translation for go through all the files phase by phase, instead of just going through all phases file by file.